### PR TITLE
ignore secrets automatically

### PR DESCRIPTION
**Pull request by MosheEichler**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/28995

## Description
added an option to ignore secrets automatically when creating a pack using `init` command in **demisto-sdk**.

## Screenshots
Paste here any images that will help the reviewer

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: demisto_sdk/commands/init/initiator.py [hunk 1]
```diff
@@ -378,6 +380,31 @@ def get_valid_user_input(options_list: List[str], option_message: str) -> str:
             except ValueError:
                 user_input = input("\nThe option must be number, please enter valid choice: ")
 
+    def ignore_secrets_automatically(self):
+        pack_dir = get_pack_name(self.full_output_path)
+        files_and_directories = glob.glob(f'{self.full_output_path}/**/*', recursive=True)
+
+        sv = SecretsValidator(white_list_path='./Tests/secrets_white_list.json')
+        # remove directories and irrelevant files
+        files = [f for f in files_and_directories if os.path.isfile(f) and sv.is_text_file(f)]
```
Comment (reutshal) [inline demisto_sdk/commands/init/initiator.py:389]: Please replace the 'f' with a proper name

File: demisto_sdk/commands/init/initiator.py [hunk 2]
```diff
@@ -378,6 +380,31 @@ def get_valid_user_input(options_list: List[str], option_message: str) -> str:
...
+        all_secrets = sv.search_potential_secrets(files)  # -> {'a': {'b': ['c', 'd'], 'e': ['f']}, 'g': ['h']}
```
Comment (reutshal) [inline demisto_sdk/commands/init/initiator.py:390]: What does this comment mean?

File: demisto_sdk/commands/init/initiator.py [hunk 3]
```diff
@@ -378,6 +380,31 @@ def get_valid_user_input(options_list: List[str], option_message: str) -> str:
...
+        secrets: set = set()
+
+        def extract_values_from_nested_dict_to_a_set(d, s):
```
Comment (reutshal) [inline demisto_sdk/commands/init/initiator.py:393]: Please move this method outside the function, provide a docstring with examples and provide better arguments and variables names.
v/d/i/s are not informative and are hard to follow.

File: demisto_sdk/commands/init/initiator.py [hunk 4]
```diff
@@ -413,6 +440,11 @@ def integration_init(self) -> bool:
         self.copy_common_server_python()
         self.copy_demistotmock()
 
+        if self.template != self.DEFAULT_INTEGRATION_TEMPLATE:  # DEFAULT_INTEGRATION_TEMPLATE there are no secrets
```
Comment (reutshal) [inline demisto_sdk/commands/init/initiator.py:443]: Did you verify that?

File: demisto_sdk/commands/init/tests/initiator_test.py [hunk 5]
```diff
@@ -470,7 +473,67 @@ def test_template_integration_init(initiator, tmpdir, template):
     assert not diff, f'There\'s a missing file in the copied files, diff is {diff}'
 
 
-def test_script_init(initiator, tmpdir):
+def test_template_integration_init_with_ignore_secrets(initiator, tmpdir, mocker):
+    """
+    Tests `integration_init` function with a given integration template name.
+
+    Given
+        - Inputs to init integration in a given output.
```
Comment (reutshal) [inline demisto_sdk/commands/init/tests/initiator_test.py:481]: This line is not clear to me, can you please rephrase it?

File: demisto_sdk/commands/init/tests/initiator_test.py [hunk 6]
```diff
@@ -470,7 +473,67 @@ def test_template_integration_init(initiator, tmpdir, template):
...
+        - An integration template - HelloWorld.
+
+    When
+        - Running the init command with secrete ignore.
+
+    Then
+        - Ensure the function's return value is True
```
Comment (reutshal) [inline demisto_sdk/commands/init/tests/initiator_test.py:488]: Which function?

File: demisto_sdk/commands/init/tests/initiator_test.py [hunk 7]
```diff
@@ -470,7 +473,67 @@ def test_template_integration_init(initiator, tmpdir, template):
...
+    expected_files = {
+        "Pipfile", "Pipfile.lock", "README.md", f"{INTEGRATION_NAME}.py",
+        f"{INTEGRATION_NAME}.yml", f"{INTEGRATION_NAME}_description.md", f"{INTEGRATION_NAME}_test.py",
+        f"{INTEGRATION_NAME}_image.png", "test_data", "command_examples"
+    }
+
+    assert res
+    assert os.path.isdir(integration_path)
+    diff = expected_files.difference(integration_dir_files)
+    assert not diff, f'There\'s a missing file in the copied files, diff is {diff}'
```
Comment (reutshal) [inline demisto_sdk/commands/init/tests/initiator_test.py:530]: Not sure this error is informative enough, what copied files? do you mean the files created by the init command?

File: demisto_sdk/commands/init/tests/initiator_test.py [hunk 8]
```diff
@@ -470,7 +473,67 @@ def test_template_integration_init(initiator, tmpdir, template):
...
+    assert os.stat(secrets_ignore_path).st_size > 0
+    shutil.rmtree('Packs')
+    shutil.rmtree('Tests')
+
+
+def test_script_init(mocker, initiator, tmpdir):
```
Comment (reutshal) [inline demisto_sdk/commands/init/tests/initiator_test.py:536]: Why aren't we adding a test for the script init?

**Review by reutshal — CHANGES_REQUESTED** [review]

**Comment by reutshal** [comment]
@MosheEichler Please add this change to the changlog

File: demisto_sdk/commands/init/tests/initiator_test.py (hunk 7 above)
Comment (reutshal) [reply demisto_sdk/commands/init/tests/initiator_test.py:528]: I dont see this has changed.
Also, I'm not sure when the user will get here.
They are just running the command, why would there be missing files?

**Review by reutshal — CHANGES_REQUESTED** [review]

File: demisto_sdk/commands/init/initiator.py [hunk 9]
```diff
@@ -25,6 +25,26 @@
 from demisto_sdk.commands.secrets.secrets import SecretsValidator
 
 
+def extract_values_from_nested_dict_to_a_set(given_dictionary: dict, return_set: set):
```
[the remark this answers is not in the document]
Comment (MosheEichler) [reply demisto_sdk/commands/init/initiator.py:28]: We can't create it inside the function, because the function is recursive, every call of the function will create a new set

File: demisto_sdk/commands/init/tests/initiator_test.py [hunk 10]
```diff
@@ -614,5 +608,3 @@ def test_script_init_with_ignore_secrets(mocker, monkeypatch, initiator, tmpdir)
     assert {f"{SCRIPT_NAME}.py", f"{SCRIPT_NAME}.yml", f"{SCRIPT_NAME}_test.py",
             "README.md", "test_data"} == script_dir_files
     assert os.stat(secrets_ignore_path).st_size > 0
-    shutil.rmtree('Packs')
```
Comment (reutshal) [inline demisto_sdk/commands/init/tests/initiator_test.py]: Why did you remove these lines?

File: demisto_sdk/commands/init/tests/initiator_test.py [hunk 11]
```diff
@@ -473,6 +475,60 @@ def test_template_integration_init(monkeypatch, initiator, tmpdir, template):
     assert not diff, f'There\'s a missing file in the copied files, diff is {diff}'
 
 
+def test_integration_init_with_ignore_secrets(initiator, tmpdir, monkeypatch, mocker):
+    """
+    Tests `integration_init` function with a given script template name.
+    Given
+        - An Integration template - HelloWorld.
```
Comment (reutshal) [inline demisto_sdk/commands/init/tests/initiator_test.py:482]: Did you test with the other templates?

File: CHANGELOG.md [hunk 12]
```diff
@@ -1,4 +1,5 @@
 # Changelog
+* Added an option to ignore secrets automatically when using the **init** command to create a pack
```
Comment (reutshal) [inline CHANGELOG.md:2]:
```suggestion
* Added an option to ignore secrets automatically when using the **init** command to create a pack.
```

File: demisto_sdk/commands/init/initiator.py [hunk 13]
```diff
@@ -19,8 +20,30 @@
...
                                                get_common_server_path,
-                                               print_error, print_v,
-                                               print_warning)
+                                               get_pack_name, print_error,
+                                               print_v, print_warning)
+from demisto_sdk.commands.secrets.secrets import SecretsValidator
+
+
+def extract_values_from_nested_dict_to_a_set(given_dictionary: dict, return_set: set):
+    """A recursive function that extracts values (the values are of type list) from a nested dictionary to a set.
```
Comment (reutshal) [inline demisto_sdk/commands/init/initiator.py:29]:
```suggestion
    """Recursively extracts values from a nested dictionary to a set.
```

File: demisto_sdk/commands/init/initiator.py [hunk 14]
```diff
@@ -19,8 +20,30 @@
...
+
+    Args:
+        given_dictionary: The nested dictionary to extract the values from.
+        return_set: the set with the return values.
```
Comment (reutshal) [inline demisto_sdk/commands/init/initiator.py:33]:
```suggestion
        return_set: the set with the extracted values.
```

File: demisto_sdk/commands/init/initiator.py [hunk 15]
```diff
@@ -19,8 +20,30 @@
...
+
+    Returns:
+        A set with the values from the dictionary.
```
Comment (reutshal) [inline demisto_sdk/commands/init/initiator.py:36]: This method doesn't seem to return anything, just adding values to the set

File: demisto_sdk/commands/init/initiator.py [hunk 16]
```diff
@@ -19,8 +20,30 @@
...
+
+    {'a': {'b': ['secret1', 'secret2'], 'e': ['secret1']}, 'g': ['secret3']} -> (secret1, secret2, secret3)
```
Comment (reutshal) [inline demisto_sdk/commands/init/initiator.py:38]: This example seems wrong since this method doesn't return anything

File: demisto_sdk/commands/init/initiator.py [hunk 17]
```diff
@@ -378,6 +401,25 @@ def get_valid_user_input(options_list: List[str], option_message: str) -> str:
...
                 user_input = input("\nThe option must be number, please enter valid choice: ")
 
+    def ignore_secrets_automatically(self):
+        pack_dir = get_pack_name(self.full_output_path)
+        files_and_directories = glob.glob(f'{self.full_output_path}/**/*', recursive=True)
+
+        sv = SecretsValidator(white_list_path='./Tests/secrets_white_list.json')
+        # remove directories and irrelevant files
+        files = [file for file in files_and_directories if os.path.isfile(file) and sv.is_text_file(file)]
+        # The search_potential_secrets method returns a nested dict with values of type list the values are the secrets
```
Comment (reutshal) [inline demisto_sdk/commands/init/initiator.py:411]:
```suggestion
        # The search_potential_secrets method returns a nested dict with values of type list. The values are the secrets
```

File: demisto_sdk/commands/init/tests/initiator_test.py [hunk 18]
```diff
@@ -473,6 +475,60 @@ def test_template_integration_init(monkeypatch, initiator, tmpdir, template):
...
+    initiator.dir_name = INTEGRATION_NAME
+    initiator.is_integration = True
+    initiator.template = 'HelloWorld'
+    initiator.category = 'Utilities'
+    mocker.patch('demisto_sdk.commands.init.initiator.get_pack_name', return_value='PackName')
+
+    integration_path = os.path.join(temp_pack_dir, INTEGRATION_NAME)
+
+    with ChangeCWD(tmpdir):
+        res = initiator.integration_init()
```
Comment (reutshal) [inline demisto_sdk/commands/init/tests/initiator_test.py:516]: Where here we are running the init with secrets ignore? Isn't that something the user should insert?

File: demisto_sdk/commands/init/tests/initiator_test.py [hunk 19]
```diff
@@ -504,3 +560,51 @@ def test_script_init(monkeypatch, initiator, tmpdir):
...
+    with ChangeCWD(tmpdir):
+        res = initiator.script_init()
+
+    script_dir_files = {file for file in listdir(script_path)}
+
+    assert res
+    assert os.path.isdir(script_path)
+    assert {f"{SCRIPT_NAME}.py", f"{SCRIPT_NAME}.yml", f"{SCRIPT_NAME}_test.py",
+            "README.md", "test_data"} == script_dir_files
+    assert os.stat(secrets_ignore_path).st_size > 0
```
Comment (reutshal) [inline demisto_sdk/commands/init/tests/initiator_test.py:610]: Why aren't we testing here the files diff like in the integration test?

**Review by reutshal — CHANGES_REQUESTED** [review]

File: demisto_sdk/commands/init/tests/initiator_test.py (hunk 10 above)
Comment (MosheEichler) [reply demisto_sdk/commands/init/tests/initiator_test.py]: because I changed that I create the files in a temporary  repo (with ChangeCWD(tmpdir)) that removes the files automatically.

File: demisto_sdk/commands/init/tests/initiator_test.py (hunk 11 above)
Comment (reutshal) [reply demisto_sdk/commands/init/tests/initiator_test.py:482]: I suggest adding at least one more template just to make sure everything is working as expected.

File: demisto_sdk/commands/init/tests/initiator_test.py (hunk 19 above)
Comment (reutshal) [reply demisto_sdk/commands/init/tests/initiator_test.py:610]: Lets try to be consistent with the UT.
If you have a test for a script that is checking the same as the integration it should look the same.
Also, please pay attention when copying code from others, this should be done carefully and with a lot of attention - we should fully understand what was done there and ask ourselves whether its relevant for our case

**Review by reutshal — CHANGES_REQUESTED** [review]

File: demisto_sdk/commands/init/initiator.py [hunk 20]
```diff
@@ -378,6 +397,25 @@ def get_valid_user_input(options_list: List[str], option_message: str) -> str:
...
+        pack_dir = get_pack_name(self.full_output_path)
+        files_and_directories = glob.glob(f'{self.full_output_path}/**/*', recursive=True)
+
+        sv = SecretsValidator(white_list_path='./Tests/secrets_white_list.json')
+        # remove directories and irrelevant files
+        files = [file for file in files_and_directories if os.path.isfile(file) and sv.is_text_file(file)]
+        # The search_potential_secrets method returns a nested dict with values of type list. The values are the secrets
+        # {'a': {'b': ['secret1', 'secret2'], 'e': ['secret1']}, 'g': ['secret3']}
+        all_secrets = sv.search_potential_secrets(files)
+        secrets: set = set()
```
Comment (reutshal) [inline demisto_sdk/commands/init/initiator.py:410]: Please improve the naming here. Its not clear what is the difference between "secrets" and "all secrets"

**Review by reutshal — APPROVED** [review]

File: demisto_sdk/commands/init/initiator.py [hunk 21]
```diff
@@ -413,6 +451,12 @@ def integration_init(self) -> bool:
         self.copy_common_server_python()
         self.copy_demistotmock()
 
+        if self.template != self.DEFAULT_INTEGRATION_TEMPLATE:  # DEFAULT_INTEGRATION_TEMPLATE there are no secrets
+            ignore_secrets = input("\nCurrently, there are some secrets in the pack. "
+                                   "Would you like ignore them automatically? Y/N ").lower()
+            if ignore_secrets in ['y', 'yes']:
```
Comment (roysagi) [inline demisto_sdk/commands/init/initiator.py:457]: 2 notes:
1. do lower case on `ignore_secrets` to accept Y or YES and such.
2. Please print the detected secrets to the user - so he'd know what he is ignoring.

File: demisto_sdk/commands/init/initiator.py [hunk 22]
```diff
@@ -378,6 +397,25 @@ def get_valid_user_input(options_list: List[str], option_message: str) -> str:
             except ValueError:
                 user_input = input("\nThe option must be number, please enter valid choice: ")
 
+    def ignore_secrets_automatically(self):
+        pack_dir = get_pack_name(self.full_output_path)
+        files_and_directories = glob.glob(f'{self.full_output_path}/**/*', recursive=True)
+
+        sv = SecretsValidator(white_list_path='./Tests/secrets_white_list.json')
```
Comment (roysagi) [inline demisto_sdk/commands/init/initiator.py:404]:
```suggestion
        sv = SecretsValidator(white_list_path='./Tests/secrets_white_list.json', ignore_entropy=True)
```
